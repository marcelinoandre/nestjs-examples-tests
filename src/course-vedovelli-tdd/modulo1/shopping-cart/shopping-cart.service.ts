import { Injectable } from '@nestjs/common';
import { find, remove } from 'lodash';
import * as Money from 'dinero.js';

// Dinero.defaultCurrency = 'BRL';
// Dinero.defaultPrecision = 2;
@Injectable()
export class ShoppingCartService {
  items = [];
  money = Money;
  constructor() {
    this.money.defaultCurrency = 'BRL';
    this.money.defaultPrecision = 2;
  }

  getTotal() {
    return this.items.reduce((acc, item) => {
      const amount = this.money({ amount: item.quantity * item.product.price });
      let discount = this.money({ amount: 0 });

      if (
        item.condition &&
        item.condition.percentage &&
        item.quantity > item.condition.minimum
      ) {
        discount = amount.percentage(item.condition.percentage);
      }

      return acc.add(amount).subtract(discount);
    }, this.money({ amount: 0 }));
  }

  add(item) {
    const itemToFind = { product: item.product };

    if (find(this.items, itemToFind)) {
      remove(this.items, itemToFind);
    }
    this.items.push(item);
  }

  remove(product) {
    remove(this.items, { product });
  }

  summary() {
    const total = this.getTotal().getAmount();
    const items = this.items;

    return {
      total,
      items,
    };
  }

  checkout() {
    const { total, items } = this.summary();

    this.items = [];

    return {
      total,
      items,
    };
  }
}
