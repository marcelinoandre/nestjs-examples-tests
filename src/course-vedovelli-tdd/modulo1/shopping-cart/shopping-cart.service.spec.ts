import { Test, TestingModule } from '@nestjs/testing';
import { ShoppingCartService } from './shopping-cart.service';

describe('ShoppingCartService', () => {
  let service: ShoppingCartService;
  const product = {
    title: 'Adidas running shows - men',
    price: 35388, //353.88 | R$ 353,83
  };

  const product2 = {
    title: 'Mizuno running shows - women',
    price: 41872,
  };

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [ShoppingCartService],
    }).compile();

    service = module.get<ShoppingCartService>(ShoppingCartService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should return 0 when getTotal() is executed in a newly created instance', () => {
    expect(service.getTotal()).toBe(0);
  });

  it('should multiply quantity and price and receive the total amount', () => {
    const item = {
      product,
      quantity: 2, // 70776
    };
    service.add(item);

    expect(service.getTotal()).toEqual(70776);
  });

  it('should ensure no more than on product exists at a time', () => {
    service.add({
      product,
      quantity: 2,
    });

    service.add({
      product,
      quantity: 1,
    });

    expect(service.getTotal()).toBe(35388);
  });

  it('should update total when a product gets included and then removed', () => {
    service.add({
      product,
      quantity: 2,
    });

    service.add({
      product: product2,
      quantity: 1,
    });

    service.remove(product);

    expect(service.getTotal()).toEqual(41872);
  });
});
