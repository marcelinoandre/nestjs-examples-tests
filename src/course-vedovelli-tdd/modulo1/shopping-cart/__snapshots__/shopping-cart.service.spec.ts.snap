// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingCartService checkout() should return an object with the total and the list of items when summary() is called. 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running shows - men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Mizuno running shows - women",
      },
      "quantity": 3,
    },
  ],
  "total": 196392,
}
`;

exports[`ShoppingCartService checkout() should return an object with the total and the list of items. 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running shows - men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Mizuno running shows - women",
      },
      "quantity": 3,
    },
  ],
  "total": 196392,
}
`;
